### PR TITLE
Moved creation of boundary conditions array out of Flow

### PR DESCRIPTION
**Pull request by b-fg**
I also tried using the lambda function `f` to adapt the arrays, but [Flow.jl:68](https://github.com/weymouth/WaterLily.jl/compare/CUDA...b-fg:CUDA?expand=1#diff-fdc9d5cd4620e715df00dfa0ce6683fc670f49831b1e589bfc4288300b546eefR68) throws an error when doing so.  It makes little sense since all the other CUDA arrays are created without any issues when using this method...
When using `adapt!` everything works as expected.

### CODE REVIEW

File: src/util.jl [hunk 1]
```diff
@@ -148,6 +148,28 @@ function slice(dims::NTuple{N}, i, j, low = 1, trim = 0) where N
...
+#         slice_donor_start = slice_ghost_start .+ δ(d, D)
+#         slice_ghost_end = slice(Ng, Ng[d] - 1, d, 1, 2)
+#         slice_donor_end = slice_ghost_end .- δ(d, D)
+#         push!(bc_list, zip(slice_ghost_start, slice_donor_start, ntuple(x -> d, length(slice_ghost_start)))...,
+#             zip(slice_ghost_end, slice_donor_end, ntuple(x -> d, length(slice_ghost_end)))...)
+#     end
+#     return Tuple.(bc_list)
+# end
+
 """
```
Comment (weymouth) [inline src/util.jl:173]: Let's uncomment this and start testing it with the test.jl file.

File: CUDAEnv/Flow.jl [hunk 2]
```diff
@@ -39,23 +53,14 @@ struct Flow{D, V, S, F, B, T}
     U :: NTuple{D, T}  # domain boundary values
     Δt:: Vector{T}  # time step
     ν :: T                  # kinematic viscosity
-    function Flow(N::NTuple{D}, U; Δt=0.25, ν=0., uλ::Function=(i, x) -> 0., T=Float64) where D
+    function Flow(N::NTuple{D}, U; Δt=0.25, ν=0., uλ::Function=(i, x) -> 0., f=identity, T=Float64) where D
```
Comment (weymouth) [inline CUDAEnv/Flow.jl:56]: Don't you need to change all the `adapt!` calls below to `f` calls?

File: src/util.jl [hunk 3]
```diff
@@ -148,6 +148,28 @@ function slice(dims::NTuple{N}, i, j, low = 1, trim = 0) where N
...
+# Vector of Tuple(s) in which each Tuple is composed of a ghost cell CartesianIndex,
+# its respective donor cell CartesianIndex, and the normal direction between them:
+# [Tuple{CartesianIndex, CartesianIndex, Int}, ...]
+# """
+# function bc_indices(Ng)
+#     D = length(Ng)
+#     bc_list = Tuple{CartesianIndex, CartesianIndex, Int}[]
+#     for d ∈ 1:D
+#         slice_ghost_start = slice(Ng, 0, d, 1, 2)
+#         slice_donor_start = slice_ghost_start .+ δ(d, D)
```
Comment (weymouth) [inline src/util.jl:164]: Is this correct for both scalars and vectors?

File: CUDAEnv/Flow.jl (hunk 2 above)
Comment (b-fg) [reply CUDAEnv/Flow.jl:56]: Please see the PR comment on why I am not using it yet.

File: src/util.jl (hunk 3 above)
Comment (b-fg) [reply src/util.jl:164]: This just gets you the list of indices for ghost-donor cells, which should be the same for both scalar and vectors AFAIK.
Then the imposition of the boundary conditions is specialized for each of them in `function BC!(u, U, bc, f = 1.0)` and `BC!(u, bc)`.

File: CUDAEnv/Flow.jl (hunk 2 above)
Comment (weymouth) [reply CUDAEnv/Flow.jl:56]: Sorry, I misunderstood the comment.

File: src/util.jl (hunk 1 above)
Comment (b-fg) [reply src/util.jl:173]: Sure. Where is that file btw?

File: src/util.jl (hunk 3 above)
Comment (weymouth) [reply src/util.jl:164]: Yes, that's true.

File: src/util.jl (hunk 1 above)
Comment (weymouth) [reply src/util.jl:173]: test\test.jl

That's the file which is run automatically when pushing up the changes to github. We should also be able to set-up GPU checks for that system.

Comment (b-fg) [reply src/util.jl:173]: You mean test/runtest.jl?

**Comment by b-fg** [comment]
Please have a look at the new commit and let me know if something is unclear.
